wasm/src/main/java/com/dylibso/chicory/wasm/io/BufferWasmInputStream.java

```diff
@@ -0,0 +1,88 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream backed by a byte buffer.
+ */
+class BufferWasmInputStream extends WasmInputStream {
+    private final ByteBuffer buffer;
+
+    BufferWasmInputStream(final ByteBuffer buffer) {
+        super();
+        // make our position independent, and zero-based
+        this.buffer = buffer.slice();
+        this.buffer.order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    public long position() {
+        return buffer.position();
+    }
+
+    public void skip(final long count) {
+        if (count <= 0) {
+            return;
+        }
+        long newOffs = buffer.position() + count;
+        if (newOffs > buffer.limit()) {
+            throw new WasmEOFException();
+        }
+        buffer.position((int) newOffs);
+    }
+
+    public int rawByteOpt() {
+        return buffer.hasRemaining() ? Byte.toUnsignedInt(buffer.get()) : -1;
+    }
+
+    public int peekRawByteOpt() {
+        return buffer.hasRemaining() ? Byte.toUnsignedInt(buffer.get(buffer.position())) : -1;
+    }
+
+    public byte[] rawBytes(final int len) {
+        byte[] buf = new byte[len];
+        try {
+            buffer.get(buf);
+        } catch (BufferUnderflowException ignored) {
+            throw new WasmEOFException();
+        }
+        return buf;
+    }
+
+    public void rawBytes(final ByteBuffer buf) {
+        int srcPos = buffer.position();
+        int srcLim = buffer.limit();
+        int dstPos = buf.position();
+        int dstLim = buf.limit();
+        int copySize = dstLim - dstPos;
+        if (copySize <= srcLim - srcPos) {
+            // reduce limit to size of copy
+            buffer.limit(srcPos + copySize);
```

Comment:
I struggle to grasp this line, and the comment further confuses me. Would you mind expanding?

Reply:
The documentation for `ByteBuffer` and `ByteBuffer#put(ByteBuffer)` explains clearly the requirements for how position and limit works. Essentially the source buffer has to be limited in size in order to avoid overflowing the target buffer, but we also must restore the original limit afterwards.

If `ByteBuffer` is confusing you, you might find [this guide](https://docs.jboss.org/xnio/2.x-userguide/html/buffers.html) helpful which I wrote for XNIO many years ago. Just ignore the XNIO parts.

Reply:
The more I go through those details the more I would love to compare an implementation based on a higher level API such as `ByteArrayInputStream` that already implements those details.

Reply:
You cannot use ByteArrayInputStream to wrap a ByteBuffer, so I fail to see how it is possible to use that class for this case.

Reply:
No, it will give an error on read-only buffers and direct (off-heap) buffers, the latter of which being part of the use case of this implementation.